MayssaraOmar/assignment2/Assignment 2.cpp

```diff
@@ -0,0 +1,275 @@
+/* ASSUMPTIONS
+ * print_ancestors method: Binary Tree has unique keys
+ * common_ancestor method: if a node is an ancestor of the other node, return that node // that way there is always an answer
+ */
+#include <iostream>
+#include <stack>
+#include <utility>
+using namespace std;
+
+template<class T>
+class binary_tree;
+
+template<class T>
+class tree_node {
+	T data;
+	tree_node *left, *right;
+public:
+	tree_node(const T &data, tree_node* left, tree_node* right) :
+			data(data), left(left), right(right) {
+	}
+	friend class binary_tree<T> ;
+// the binary_tree class is declared as tree_node's friend which means that binary_tree can access the private members of tree_node
+};
+
+template<class T>
+class binary_tree {
+
+	tree_node<T> *root;
+public:
+	binary_tree();
+	~binary_tree();
+	tree_node<T>* insert_root(const T &);
+	tree_node<T>* insert_left(tree_node<T>*, const T &); // insert an element to the left of the node // if an element to the left already exists, it overwrites it
+	tree_node<T>* insert_right(tree_node<T>*, const T &); // insert an element to the right of the node // if an element to the right already exists, it overwrites it
+	tree_node<T>* get_root();
+	tree_node<T>* get_left(tree_node<T>*);
+	tree_node<T>* get_right(tree_node<T>*);
+	void remove_leaf(tree_node<T>*); // removes a leaf node // if the node isn't a leaf it doesn't remove anything
+	bool empty();
+	void clear();
+	void print_ancestors(const T &target); // print ancestors' keys starting from the parent to the root separated by spaces
+	pair<tree_node<T>*, T> common_ancestor(tree_node<T>*, tree_node<T>*); // in case of an error, the first part of the returned pair will be nullptr
+private:
+	bool print_ancestors_helper(tree_node<T>* current, const T &target);
+	pair<tree_node<T>*, int> common_ancestor_helper(tree_node<T>*,
+			tree_node<T>*, tree_node<T>*);
+
+};
+template<class T>
+binary_tree<T>::binary_tree() {
+	root = nullptr;
+}
+
+template<class T>
+binary_tree<T>::~binary_tree() {
+	clear();
+}
+
+template<class T>
+tree_node<T>* binary_tree<T>::insert_root(const T &data) {
+	return root = new tree_node<T>(data, nullptr, nullptr);
+}
+
+template<class T>
+tree_node<T>* binary_tree<T>::insert_left(tree_node<T>* current,
```

Comment:
It's a bit strange taking a node* as parameter, for example, it could be a node from another tree.
You could verify here that the tree doesn't already have a node with the same data.